geode-core/src/test/java/org/apache/geode/internal/cache/CacheEventListenerDUnitTest.java

```diff
@@ -0,0 +1,541 @@
+import org.apache.geode.cache.EntryEvent;
+import org.apache.geode.cache.ExpirationAction;
+import org.apache.geode.cache.ExpirationAttributes;
+import org.apache.geode.cache.Operation;
+import org.apache.geode.cache.Region;
+import org.apache.geode.cache.RegionEvent;
+import org.apache.geode.cache.Scope;
+import org.apache.geode.cache.util.CacheListenerAdapter;
+import org.apache.geode.internal.logging.LogService;
+import org.apache.geode.test.dunit.Host;
+import org.apache.geode.test.dunit.Invoke;
+import org.apache.geode.test.dunit.cache.rules.CacheRule;
+import org.apache.geode.test.dunit.rules.DistributedTestRule;
+import org.apache.geode.test.dunit.rules.SharedCountersRule;
+import org.apache.geode.test.junit.categories.DistributedTest;
+import org.apache.geode.test.junit.rules.serializable.SerializableTestName;
+
+/**
+ * @since 2.0
+ */
+@Category(DistributedTest.class)
+@SuppressWarnings("serial")
+public class CacheEventListenerDUnitTest implements Serializable {
+
+  private static final Logger logger = LogService.getLogger();
+
+  private static final String REGION_NAME = CacheEventListenerDUnitTest.class.getSimpleName();
+  private static final Scope SCOPE = Scope.DISTRIBUTED_ACK;
+
+  private static final String NUM_INVOCATIONS_1 = "NUM_INVOCATIONS_1";
+  private static final String NUM_INVOCATIONS_2 = "NUM_INVOCATIONS_2";
+
+  private volatile static Object newValue;
+  private volatile static Object oldValue;
+
+  private String name;
+  private Host host;
+  private int numVMs;
+
+  @ClassRule
+  public static DistributedTestRule distributedTestRule = new DistributedTestRule();
+
+  @Rule
+  public SharedCountersRule sharedCountersRule = SharedCountersRule.builder().build();
+
+  @Rule
+  public SerializableTestName testName = new SerializableTestName();
+
+  @Rule
+  public CacheRule cacheRule = CacheRule.builder().createCacheInAll().build();
+
+  @Before
+  public void setUp() throws Exception {
+    name = testName.getMethodName();
+
+    host = Host.getHost(0);
+    numVMs = host.getVMCount() + 1;
+
+    sharedCountersRule.initialize(NUM_INVOCATIONS_1);
+    sharedCountersRule.initialize(NUM_INVOCATIONS_2);
+
+    cacheRule.getCache().createRegionFactory().setScope(SCOPE).create("root");
+    Invoke.invokeInEveryVM(() -> {
+      cacheRule.getCache().createRegionFactory().setScope(SCOPE).create("root");
+    });
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    newValue = null;
+    oldValue = null;
+    Invoke.invokeInEveryVM(() -> {
+      newValue = null;
+      oldValue = null;
+    });
+  }
+
+  @Test
+  public void testObjectAddedReplaced() throws Exception {
+    CacheEventListener listener = new CacheEventListener() {
+
+      @Override
+      public void afterCreate(EntryEvent event) {
+        logger.info("Invoking afterCreate on listener; name={}", event.getKey());
+
+        assertThat(event.getDistributedMember()).isEqualTo(event.getCallbackArgument());
```

Comment:
All of these assertions in the callbacks need to be changed to use ErrorCollector JUnit Rule or something similar.